docs/architecture/adr-071-cryptography-v2.md

```diff
@@ -0,0 +1,566 @@
+
+Plugin Architecture and Dependency Injection
+
+* Establish the architectural foundation for an extensible plugin system and integrate a dependency injection framework, ensuring modularity, testability, and third-party integrations.
+
+* Design an environment for plugin testing, ensuring developers can validate integrations without compromising system integrity.
+
+Interface considerations
+
+* Design should take into considerations support for Trusted Platform Module (TPM) 2.0 and similar devices to anticipate future enhancements.
+
+* Design should take into account the Cryptographic Token Interface Standard (PKCS#11)
+
+Increase cryptographic versatility
+
+* Support for a broad spectrum of cryptographic techniques
+* Extend support for more hash functions (e.g. pedersen, argon2, Argon2d/I/id, Blake3, etc.)
+* Extend support for more signature schemes (e.g. secp256r1, ed25519, ed448, sr25519, etc.)
+* More advanced methods ( Post-Quantum Cryptography (PQC) methods
+* Threshold signatures and encryption
+
+Community Engagement Infrastructure:
+
+* Structure the design with tools and documentation interfaces in mind, enabling a seamless future rollout of resources for developer engagement.
+
+## Proposed architecture
+
+### Introduction
+
+In the proposed architecture, each package is decoupled and isolated. Adding new implementations consist of implementing the required interfaces.
+
+```mermaid
+classDiagram
+
+Keyring <|-- Wallet
+
+SecureStorage <|-- Keyring
+SecureItem <|-- SecureStorage
+CryptoProvider <|-- SecureItem
+
+Hasher <|-- CryptoProvider
+Verifier <|-- CryptoProvider
+Signer <|-- CryptoProvider
+Cipher <|-- CryptoProvider
+Generator <|-- CryptoProvider
+```
+
+### Crypto Provider
+
+*Crypto Providers* serve as a middleware responsible for managing the interaction with various instantiated cryptographic packages. It acts as a centralized controller, encapsulating the API of the crypto modules in a single location.
+Through each Crypto provider, users can access functionality such as signing, verification, encryption, and hashing.
+
+By abstracting the underlying cryptographic functionality, *Crypto providers* enable a modular and extensible architecture. It allows users to easily switch between different cryptographic implementations without impacting the rest of the system.
+
+A `CryptoProvider` can be built from different sources, such as a `SecureItem`, a randomness source, a seed, a mnemonic, or a string.
+Since the underlying representation of a `CryptoProvider` is a protobuf message, it can be easily serialized and stored in a safe manner.
+The storing and retrieval is handled by a `SecureStorage` implementation which will be presented in the next section.
+
+```go
+type ProviderMetadata interface {
+  key string
+  value string
+}
+
+type ICryptoProviderMetadata interface {
+  GetTypeUUID() TypeUUID
+  GetName() string
+  GetMetadata() []ProviderMetadata
+}
+
+type ICryptoProviderBuilder interface {
```

Comment:
Are these interfaces meant to be implemented by modules outside the SDK? If so, you should consider that you can't add methods without breaking backwards compatibility. That means `ICryptoProviderBuilder` and `ICryptoProvider` are not extensible to other operations or sources that don't fit one of the existing methods.

Reply:
`ICryptoProviderBuilder` would likely not change that much since it's the provider factory, shouldn't involve crypto specifics in its code. 
And regarding `ICryptoProvider`, all the possible getters (GetCipher, GetHasher, etc) should be defined in this interface. 
What getter you think we should add to cover more backends?